peer-ids/peer-ids.md

```diff
@@ -0,0 +1,80 @@
+# Spec:  Peer Ids and Keys
+
+## Keys
+
+Keys are serialized for transmission using a
+[simple protobuf](https://github.com/libp2p/go-libp2p-crypto/blob/master/pb/crypto.proto#L5):
+
+```protobuf
+enum KeyType {
+	RSA = 0;
+	Ed25519 = 1;
+	Secp256k1 = 2;
+	ECDSA = 3;
+}
+
+message PublicKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+```
+
+This proto simply encodes for transmission a public key along with an enum specifying the type of key. The specific format of the `Data` field depends on the key type, and is [described below](#how-keys-are-encoded-and-messages-signed).
+
+#### Where it's used?
+
+Keys are used in two places in libp2p.  The first is for signing messages.  Here are some examples of messages we sign:
+ - IPNS records
+ - PubSub messages (coming soon)
+ - SECIO handshake
+
+The second is for generating peer ids; this is discussed in the section below.
+
+## Peer Ids
+
+Here is the process by which we generate peer ids based on the public keys described above:
+
+  1. Encode the public key into the protobuf.
+  2. Serialize the protobuf containing the public key into bytes using the [canonical protobuf encoding](https://developers.google.com/protocol-buffers/docs/encoding).
```

Comment:
`1=0,2=bytes` and `2=bytes,1=0` are two equally valid encodings for an RSA key. Indeed, so is `1=2,2=bytes,1=0` given that the last value overrides the first, per the protobuf spec. Protobuf is not deterministic in its encoding - this would be a problem in general.

Reply:
Hmm, that's true. From the linked doc:

> When a message is serialized, there is no guaranteed order for how its known or unknown fields should be written. 

Are we doing anything to enforce deterministic field ordering? If not, maybe we should be...

Reply:
Some notes about the lack of deterministic serialization for protobuf: https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510

This is a thorny issue... a libp2p implementor could do everything according to the spec and still end up with peer ids that other peers wouldn't validate, if their protobuf implementation happened to order the keys differently.

Reply:
It helps to think of a protobuf message as a hash map, essentially - semantically it's a very similar situation - the last thing you put in remains, and the order is unspecified.

In fact, that's cheap and easy way to implement it - I'd expect dynamic languages with strong dictionary implementations to do it this way. 

Locking down field order solves the issue, but I'm not sure I'd call it protobuf any more at that point - perhaps protobuf-compatible.

Reply:
Yeah, I'm not crazy about requiring semantics on top of protobuf - at that point I'd rather just specify a different serialization that is deterministic. Then you have a backwards compatibility problem though...

@marten-seemann do you have any concerns about the non-deterministic encoding for this public key structure? Asking because it gets embedded in the certificate extension for the TLS 1.3 spec you're working on.

Reply:
@Stebalien @marten-seemann would appreciate your views here. I think @arnetheduck has surfaced a legitimate concern. Likely it hasn’t been an issue so far because of the simplicity of the protobuf, which makes it highly unlikely that serialisers will play tricks. But indeed, a lack of determinism is dangerous.

Reply:
We generally require that fields be ordered by tag. I believe all(?) implementations do this and it should usually "just work". However, this _spec_ may need to specify this.

A more general solution is to tell users to never re-serialize. That is, when you get a public key from the network, you can read it but you shouldn't _re_-serialize it when computing the peer ID. Instead, you should keep the serialized bytes around and use those. That would also solve https://github.com/libp2p/go-libp2p-crypto/issues/51.

Reply:
Alternatively, this protobuf is so simple we can just remove the dependency on protobuf and inline the encoding/decoding procedure by retrofitting the protobuf representation, enforcing the order and determinism.

Reply:
We could then state: “you MAY use a protobuf encoder to serialise the key, if and only if XYZ. Else you MUST use an alternative means to enforce XYZ in order to guarantee interoperability and universality.”

Reply:
Either way but I _would_ like to be able to add additional fields in the future (which is problematic with respect to re serializing).

Reply:
> When a message is serialized, there is no guaranteed order for how its known or unknown fields should be written.

The way I interpret this line, specially since it's explicitly stated in their spec, is that regardless of what implementations happen to doing today, there's no guarantee that they will do so tomorrow. The absolutely easiest way to implement a protobuf library in JavaScript for example is to use an object (I've done that for some custom projects) - with only 2 fields you have a pretty good 50% chance that the order will be correct. You then get to roll the dice every time a new version of chrome is out ;)

Another correct encoding of an RSA key is simply `2=bytes` - in protobuf 3 there's no distinction between a default value (0) and an explicitly set value, so implementations are effectively not required to spell it out. In protobuf2, the situation is similar, though you can detect a missing field in the generated API. They were able to do this change in the high-level language between v2 and v3 without changing the wire spec, largely because of nuances like this.

The only safe thing to say is that you can use a protobuf decoder - these will always be compatible.

I think stating that you may use a protobuf encoder would be shooting future users in the foot, specially if the aim is to create a standard that is language and implementation agnostic. This would be a concern, since we're considering a new libp2p impl in a new language (Nim).

> inline the encoding/decoding procedure

For a message like this, the key is a varint that includes a few bits for field type + field number and then a varint for the value. the second one is again typed varint, and bytes means a varint for length+the bytes themselves (afair). Only the encoding needs to be locked down (it would be a bit unusual, as far as specs go, though).

> never re-serialize

One of the reasons I've seen people use protobuf is to have the ability to re-serialize - it was a design goal of the language, and manifests itself as the "UnknownFields" map in the C++ API, for example. Prohibiting this takes the spec another step away from this being protobuf. Granted, it's likely not the first thing someone will want to do with a peer id specifically..

Reply:
> The way I interpret this line, specially since it's explicitly stated in their spec, is that regardless of what implementations happen to doing today, there's no guarantee that they will do so tomorrow.

Specifically, it says:

> Serialization order is an implementation detail and the details of any particular implementation may change in the future.

(i.e., implementation defined)

Annoyingly, they actually changed this. It [used to say](https://web.archive.org/web/20140215063318/https://developers.google.com/protocol-buffers/docs/encoding#order):

> While you can use field numbers in any order in a .proto, when a message is serialized its known fields should be written sequentially by field number, as in the provided C++, Java, and Python serialization code.

Which is why we relied on this. All fully-featured and correct implementations ensured their field orders matched.

> in protobuf 3 there's no distinction between a default value (0) and an explicitly set value

We currently require proto2 (these fields are actually marked as _required_ which doesn't exist in proto3) but yeah, implementers may ignore this.

> One of the reasons I've seen people use protobuf is to have the ability to re-serialize - it was a design goal of the language, and manifests itself as the "UnknownFields" map in the C++ API, for example. Prohibiting this takes the spec another step away from this being protobuf. Granted, it's likely not the first thing someone will want to do with a peer id specifically..

IMO, that's a design goal when _editing_. Any edits will change the peer ID anyways.

---

My worry with saying "this is not protobuf but you can use a protobuf decoder" is that people _will_ use protobuf encoders _anyways_ because they'll usually "just work". Then, when the underlying protobuf implementation changes, bad things will happen.

Reply:
@Stebalien 

> My worry with saying "this is not protobuf but you can use a protobuf decoder" is that people will use protobuf encoders anyways because they'll usually "just work". Then, when the underlying protobuf implementation changes, bad things will happen.

We can include test vectors for that. It's fair for implementors to use protobuf encoders, and I expect that to be the dominant implementation approach. However, if between version X and Y a specific encoder changed in a way that it broke our requirement, that's when the implementor would have to craft a manual encoding. It's a fair trade-off. We can provide guidance in this spec (in the form of "implementor's notes").

> https://web.archive.org/web/20140215063318/https://developers.google.com/protocol-buffers/docs/encoding#order
> Which is why we relied on this. All fully-featured and correct implementations ensured their field orders matched.

Woah, thanks for tracing that back 😓 

> Either way but I would like to be able to add additional fields in the future (which is problematic with respect to re serializing).

IIUC, you'd like nodes to tolerate unrecognised fields and use the struct's byte string verbatim when calculating the peer ID?

Isn't this dangerous? This creates a surface for polymorphic peer IDs, where node `A` with pubkey `P` could produce infinite peer ID (or craft specific ones) for itself – and hence break protocols that rely on peer IDs – merely by setting random fields in the protobuf.

Reply:
> Isn't this dangerous? This creates a surface for polymorphic peer IDs, where node A with pubkey P could produce infinite peer ID (or craft specific ones) for itself – and hence break protocols that rely on peer IDs – merely by setting random fields in the protobuf.

This should be just fine.

1. You can probably already do this. We don't make any effort to ensure that the `data` part of the key is completely deterministic.
2. Any change to the key metadata will, effectively, create a new identity. IMO, that's _exactly_ what we want.

Reply:
> We don't make any effort to ensure that the data part of the key is completely deterministic.

Could you elaborate?

> Any change to the key metadata will, effectively, create a new identity. IMO, that's exactly what we want.

That feels wrong. Identity should be derived from cryptographic material (a pubkey), not from random metadata.

Reply:
Not sure exactly what @Stebalien is referring to about the data field, but the spec describes one case where the same public key could lead to different data fields (and thus peer ids). 

Since there are two supported methods for encoding Ed25519 keys, you could end up with different peer ids for the same key depending on which you choose.